File: bindings/js/examples/search-token-router.mjs
import { Model, equationPrompt } from "../search.mjs";

const model = new Model("LLM token router");
const requests = ["chat", "code", "batch"];
const tokenUnits = [1, 2, 5]; // rounded 1k-token billing units
const routes = [];
const costs = [];
const latencies = [];
const localLoads = [];

for (const [index, request] of requests.entries()) {
  const route = model.choice(`${request}_route`, [
    "local",
    "fast_cloud",
    "cheap_cloud",
  ]);
  const cost = model.int(`${request}_cost`, 0, 20);
  const latency = model.int(`${request}_latency`, 45, 180);
  const localLoad = model.int(`${request}_local_load`, 0, tokenUnits[index]);
  model.table(
    [route, cost, latency, localLoad],
    [
      ["local", 0, 180, tokenUnits[index]],
      ["fast_cloud", 20, 45, 0],
      ["cheap_cloud", 7, 120, 0],
    ],
  );
  routes.push(route);
  costs.push(cost);
  latencies.push(latency);
  localLoads.push(localLoad);
}
model.add("chat_latency <= 100");
model.add("code_latency <= 200");
model.add("batch_latency <= 200");
model.add("chat_local_load + code_local_load + batch_local_load <= 5");
model.minimize(
  costs.map((cost, index) => `${tokenUnits[index]} * ${cost}`).join(" + "),
);

console.log(
  equationPrompt(
    [...latencies, ...localLoads],
    "chat latency must be at most 100 ms; code and batch latency must be at most 200 ms; total local load must be at most 5",
  ),
);
const result = model.solve({ timeoutMs: 2_000 });
if (result.status !== "optimal") {
  throw new Error(`router has no proved optimum: ${result.status}`);
}
const solution = result.requireSolution();
requests.forEach((request, index) =>
  console.log(`${request}: ${solution.get(routes[index])}`),
);
console.log(`proved minimum weighted cost: ${result.objective}`);

File: bindings/python/examples/search_token_router.py
"""Optimize an LLM request router with table constraints and a token budget."""

from aysearch import Model, equation_prompt


model = Model("LLM token router")
requests = ["chat", "code", "batch"]
# Rounded 1k-token billing units keep the objective directly interpretable.
token_units = [1, 2, 5]

routes = []
costs = []
latencies = []
local_loads = []
for index, request in enumerate(requests):
    route = model.choice(f"{request}_route", ["local", "fast_cloud", "cheap_cloud"])
    cost = model.int(f"{request}_cost", 0, 20)
    latency = model.int(f"{request}_latency", 45, 180)
    local_load = model.int(f"{request}_local_load", 0, token_units[index])
    # Rows are route, micro-dollars per 1k tokens, milliseconds, and local load.
    model.table(
        [route, cost, latency, local_load],
        [
            ["local", 0, 180, token_units[index]],
            ["fast_cloud", 20, 45, 0],
            ["cheap_cloud", 7, 120, 0],
        ],
    )
    routes.append(route)
    costs.append(cost)
    latencies.append(latency)
    local_loads.append(local_load)

# The strings below are exactly the sort of output an LLM can safely propose:
# they remain data and AY parses a deliberately restricted linear grammar.
model.add("chat_latency <= 100")
model.add("code_latency <= 200")
model.add("batch_latency <= 200")
model.add("chat_local_load + code_local_load + batch_local_load <= 5")
weighted_cost = " + ".join(
    f"{units} * {cost.name}" for units, cost in zip(token_units, costs)
)
model.minimize(weighted_cost)

print(
    equation_prompt(
        [*latencies, *local_loads],
        "chat latency must be at most 100 ms; code and batch latency must be at "
        "most 200 ms; total local load must be at most 5",
    )
)
result = model.solve(timeout_ms=2_000)
if result.status != "optimal":
    raise RuntimeError(f"router has no proved optimum: {result.status}")
solution = result.require_solution()
for request, route in zip(requests, routes):
    print(f"{request:5}: {solution[route]}")
print(f"proved minimum weighted cost: {result.objective}")
